htdocs/PI/write/PIWriteRequest.php

```diff
@@ -904,6 +806,155 @@ private function updateSiteExtensionPropertiesDelete (Site $siteService, $site,
     }
+
+  /**
+   * Updates the specified service
+   * @param  Site   $siteService [description]
+   * @throws \Exception
+   */
+  private function updateService(Site $siteService){
+    #This case requires the serviceService
+    $this->checkServiceServiceSet();
+
+    #Identify the service
+    try {
+      $service = $this->serviceService->getService($this->entityID);
+    } catch (\Exception $e) {
+      $this->httpResponseCode=404;
+      throw new \Exception("A service with the specified id could not be found");
+    }
+
+    #Authorisation
+    try {
+      $siteService->checkAuthroisedAPIIDentifier($service->getParentSite(), $this->userIdentifier, $this->userIdentifierType);
+    } catch(\Exception $e){
+      $this->httpResponseCode=403;
+      throw $e;
+    }
+
+    switch($this->entityProperty) {
+      case 'extensionproperties':{
+        $this->updateServiceExtensionProperties($service);
+        break;
+      }
+      default: {
+        $this->httpResponseCode=501;
+        throw new \exception($this->genericExceptionMessages["entityTypePropertyCombo"]);
+      }
+    }
+  }
+
+  /**
+   * Updates the extension properties of the service specified
+   *
+   * @param  $service serivce being updated
+   * @throws \Exception
+   */
+  private function updateServiceExtensionProperties($service) {
+    #TWO CASES: one there is a single value, the other mutliple - create an array for the single, use array for multiple.
+    if (is_null($this->entityPropertyKey)) {
+        $extensionPropKVArray = $this->entityPropertyKVArray;
+    } else {
+        $extensionPropKVArray = array($this->entityPropertyKey => $this->entityPropertyValue);
+    }
+
+    #Based on Request types run either an add with or without overwritebv or a remove.
+    switch ($this->requestMethod) {
+      case 'POST':{
+        $this->updateServiceExtensionPropertiesPost($service, $extensionPropKVArray);
+        break;
+      }
+      case 'PUT':{
+        $this->updateServiceExtensionPropertiesPut($service, $extensionPropKVArray);
+        break;
+      }
+      case 'DELETE':{
+          $this->updateServiceExtensionPropertiesDelete($service, $extensionPropKVArray);
+          break;
+      }
+      default: {
+        $this->httpResponseCode=405;
+        throw new \Exception($this->genericExceptionMessages["entityTypePropertyMethod"]);
+        break;
+      }
+    }
+  }
+
+  /**
+   * Update extension properties for the service requested following a POST
+   * request
+   * @param  $service service being updated
+   * @param  array $extensionPropKVArray
+   * @throws \Exception
+   */
+  private function updateServiceExtensionPropertiesPost ($service, $extensionPropKVArray) {
+    #Post requests will fail if the property with that value already exists or the property limit has been reached
+    #TODO:This will return the wrong error code if something else goes wrong
```

Comment:
Is there a plan to solve this and the other TODO?

Reply:
This was carried forward in the refactor. It could be achieved with some refactoring of the serivceService to provide some generic functions for checking these things. I don't think it is required for this PR though.